graylog2-server/src/main/java/org/graylog/plugins/views/search/elasticsearch/searchtypes/ESMessageList.java

```diff
@@ -83,8 +85,22 @@ public void doGenerateQueryPart(SearchJob job, Query query, MessageList messageL
 
         applyHighlightingIfActivated(searchSourceBuilder, job, query);
 
+        final Set<String> effectiveStreamIds = messageList.effectiveStreams().isEmpty()
+                ? query.usedStreamIds()
+                : messageList.effectiveStreams();
+
         final List<Sort> sorts = firstNonNull(messageList.sort(), Collections.singletonList(Sort.create(Message.FIELD_TIMESTAMP, SortOrder.DESC)));
-        sorts.forEach(sort -> searchSourceBuilder.sort(sort.field(), sort.order()));
+        sorts.forEach(sort -> searchSourceBuilder.sort(
+                SortBuilders.fieldSort(sort.field())
+                        .order(sort.order())
+                        .unmappedType(unmappedTypeForSort(sort, queryContext, effectiveStreamIds))
+        ));
+    }
+
+    private String unmappedTypeForSort(Sort sort, ESGeneratedQueryContext queryContext, Set<String> effectiveStreamIds) {
+        final Map<String, Set<String>> allFieldTypes = firstNonNull(queryContext.fieldTypes(effectiveStreamIds), Collections.emptyMap());
```

Comment:
I'd suggest pushing the null check down to `ESGeneratedQueryContext` or even `FieldTypesLookup`.

I even think all of the logic in this method could be pushed down to `FieldTypesLookup`. Instead of asking for all field types for these streams, ask explicitly for the type of the field, i.e., `FieldTypesLookup#getType(String fieldName, Set<String> streamIds)`. That would encapsulate all this stuff in `FieldTypesLookup` and this method could be replaced with: `queryContext.getFieldType(sort.field(), effectiveStreamIds)`

Reply:
Good idea! ✅